csrc/optimization/layout_inference.cpp

```diff
@@ -0,0 +1,236 @@
+// clang-format off
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2023-present NVIDIA CORPORATION & AFFILIATES.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+// clang-format on
+#include <ir/all_nodes.h>
+#include <ir/utils.h>
+#include <optimization/layout_inference.h>
+
+namespace nvfuser {
+
+namespace {
+
+class MemoryFormatInferencer : public OptOutConstDispatch {
+ public:
+  MemoryFormatInferencer(
+      std::unordered_map<const TensorView*, MemoryFormat>& format_map)
+      : format_map_(format_map) {}
+
+ private:
+  void handle(const UnaryOp*) override;
+  void handle(const BinaryOp*) override;
+  void handle(const BroadcastOp*) override;
+  // TODO: Add more propagation rules
+  //   void handle(const Reduction*) override;
+  //   void handle(const LoadStoreOp*) override;
+  //   void handle(const SqueezeOp*) override;
+  //   void handle(const ExpandOp*) override;
+
+ private:
+  // format_map_ records the stride order (memory format) of each TensorView.
+  // Since it only handles permutation from a rfactor domain to allocation
+  // domain, it can be interpreted as:
+  //
+  // e.g. TV0 rfactor domain [i0, i1, i2]
+  //            alloc domain [i0, i2, i1]
+  //           memory format   0,  2,  1
+  std::unordered_map<const TensorView*, MemoryFormat>& format_map_;
```

Comment:
I haven't touched that yet.

But I think it should be 3d here. i.e. we'll want to exclude reduction iterdomain, since it doesn't help `resolve` propagation with a binary op. We can probably just leave the reduction iterdomain on the left of allocation domain... or better yet, maybe we should just remove it from allocation domain since it doesn't carry any real meaning.